clients/shared/merge.go

```diff
@@ -0,0 +1,164 @@
+package shared
+
+import (
+	"fmt"
+	"log/slog"
+	"time"
+
+	"github.com/artie-labs/transfer/lib/jitter"
+
+	"github.com/artie-labs/transfer/lib/config"
+	"github.com/artie-labs/transfer/lib/config/constants"
+	"github.com/artie-labs/transfer/lib/destination"
+	"github.com/artie-labs/transfer/lib/destination/ddl"
+	"github.com/artie-labs/transfer/lib/destination/dml"
+	"github.com/artie-labs/transfer/lib/destination/types"
+	"github.com/artie-labs/transfer/lib/optimization"
+	"github.com/artie-labs/transfer/lib/ptr"
+	"github.com/artie-labs/transfer/lib/sql"
+	"github.com/artie-labs/transfer/lib/typing/columns"
+)
+
+func Merge(dwh destination.DataWarehouse, tableData *optimization.TableData, cfg config.Config, opts types.MergeOpts) error {
+	if tableData.ShouldSkipUpdate() {
+		return nil
+	}
+
+	fqName := tableData.ToFqName(dwh.Label(), true, cfg.SharedDestinationConfig.UppercaseEscapedNames, "")
+	tableConfig, err := dwh.GetTableConfig(tableData)
+	if err != nil {
+		return err
+	}
+
+	srcKeysMissing, targetKeysMissing := columns.Diff(tableData.ReadOnlyInMemoryCols(), tableConfig.Columns(),
+		tableData.TopicConfig.SoftDelete, tableData.TopicConfig.IncludeArtieUpdatedAt,
+		tableData.TopicConfig.IncludeDatabaseUpdatedAt, tableData.Mode())
+
+	createAlterTableArgs := ddl.AlterTableArgs{
+		Dwh:               dwh,
+		Tc:                tableConfig,
+		FqTableName:       fqName,
+		CreateTable:       tableConfig.CreateTable(),
+		ColumnOp:          constants.Add,
+		CdcTime:           tableData.LatestCDCTs,
+		UppercaseEscNames: &cfg.SharedDestinationConfig.UppercaseEscapedNames,
+	}
+
+	// Columns that are missing in DWH, but exist in our CDC stream.
+	err = ddl.AlterTable(createAlterTableArgs, targetKeysMissing...)
+	if err != nil {
+		slog.Warn("Failed to apply alter table", slog.Any("err", err))
+		return err
+	}
+
+	// Keys that exist in DWH, but not in our CDC stream.
+	deleteAlterTableArgs := ddl.AlterTableArgs{
+		Dwh:                    dwh,
+		Tc:                     tableConfig,
+		FqTableName:            fqName,
+		CreateTable:            false,
+		ColumnOp:               constants.Delete,
+		ContainOtherOperations: tableData.ContainOtherOperations(),
+		CdcTime:                tableData.LatestCDCTs,
+		UppercaseEscNames:      &cfg.SharedDestinationConfig.UppercaseEscapedNames,
+	}
+
+	if err = ddl.AlterTable(deleteAlterTableArgs, srcKeysMissing...); err != nil {
+		slog.Warn("Failed to apply alter table", slog.Any("err", err))
+		return err
+	}
+
+	tableConfig.AuditColumnsToDelete(srcKeysMissing)
+	tableData.MergeColumnsFromDestination(tableConfig.Columns().GetColumns()...)
+	temporaryTableName := fmt.Sprintf("%s_%s", tableData.ToFqName(dwh.Label(), false, cfg.SharedDestinationConfig.UppercaseEscapedNames, ""), tableData.TempTableSuffix())
```

Comment:
For BigQuery the temporary table name has changed from:
`fmt.Sprintf("%s_%s", tableData.RawName(), tableData.TempTableSuffix())`
To:
`temporaryTableName := fmt.Sprintf("%s_%s", tableData.ToFqName(dwh.Label(), false, cfg.SharedDestinationConfig.UppercaseEscapedNames, ""), tableData.TempTableSuffix())`

Reply:
Probably fine but calling it out.

Reply:
Looks like there's a discrepancy between PutTable: https://github.com/artie-labs/transfer/blob/8daf5a96c16cb8c3282df9c911861db309fb88ab/clients/bigquery/merge.go#L123

Reply:
Yeah `PutTable` requires us to explicitly spell out `dataset` and `projectid` while keeping the table name relative.